### PR TITLE
source switching

### PR DESCRIPTION
**Pull request by torkus**
- [x] multiple source locations are captured from the TOC file
  - can we rely on scraping this from other locations? I don't think so :(
- [x] an action exists that explicitly models the implicit process that happens when an addon is matched against a different catalogue item. 
  - we already support source switching but it's not directly under the user's control
- [x] ignored and pinned addons cannot be switched
- [x] a GUI action to search the catalogue for similarly named addons 
  - I've added a "Find similar" context menu option that does a catalogue search against the addon's `label`. 
  - It's the bare minimum but it kinda sorta *does* work - I've found curseforge addons with no alts in the wowi and github catalogues already.
- [x] GUI, add a 'source-list' column
- [ ] ~`group-id` is using the addon's `url` as a unique identifier and should be changed at the same time the source is switched~
  - should I just use a uuid? I think something in the import logic now depends on the group-id being a URL :(
    - looks like the mutual dependency checking relies on the group-id being predictable, so a unique uuid futzes with that
  - nope, this is complex enough to warrant it's own dedicated task. It's also fine for now
- [x] investigate 'synthetic' addon summaries that are built from the nfo data and the (now) incorrect group-id
  - how big of a deal is this?
    - not that big a deal. I've disabled the group-id from being used and instead recreate the URL if possible. curseforge addons now miss out on still being able to be checked if they can't be found in the catalogue, but curseforge is disappearing anyway.
- [x] case: a github addon has IDs for curseforge and wowinterface but nothing for github. switching to wowinterface would 'lose' github. 
  - nfo data should preserve a list of sources ever **used**, including the current one.
    - for example, a curseforge addon is installed, it is added to the list, but then switched to wowinterface because the addon supports it. The nfo data `source` and `source-id` are updated to wowinterface, and "wowinterface", present from the toc data, is added to the list of sources.
- [x] gui, list of releases are refreshed when an addon switches source
- [x] review
- [x] update CHANGELOG

### CODE REVIEW

**Comment by torkus** [comment]
woooooooooo - back from holiday, back from working on another project, back to working on strongbox. I feel a little guilty I've neglected it for about a week now.

File: src/strongbox/addon.clj [hunk 1]
```diff
@@ -116,18 +117,27 @@
...
     addon-list))
 
-(defn-spec ungroup-addon :addon/installed-list
-  "an addon may actually be many addons bundled together with a primary one chosen to represent them.
-  sometimes we want to treat this addon as a list of addons"
-  [addon :addon/installed]
-  (if (empty? (:group-addons addon))
-    [addon]
-    (get addon :group-addons [])))
+(defn merge-source-map-lists
```
Comment (torkus) [inline src/strongbox/addon.clj:120]: there isn't anything source-map specific about this fn. perhaps stick it in utils?

tests

defn-spec

File: src/strongbox/addon.clj [hunk 2]
```diff
@@ -116,18 +117,27 @@
...
+  "merges two lists of source-maps returning a single distinct one or nil if empty"
+  [list-a, list-b]
+  (some->> (into list-a list-b)
+           (remove nil?)
+           utils/nilable
+           distinct
+           vec))
+
+(defn extract-source-map-list
+  "extracts a source-map from the given `data` as well as anything in :source-map-list and returns a single distinct source map list"
```
Comment (torkus) [inline src/strongbox/addon.clj:130]: improve docstr

defn-spec

tests

File: src/strongbox/addon.clj [hunk 3]
```diff
@@ -116,18 +117,27 @@
...
+  [data]
+  (merge-source-map-lists
+   (or (some-> data (select-keys [:source :source-id]) utils/nilable vector) [])
+   (get data :source-map-list [])))
 
 (defn merge-toc-nfo
-  "it's own function because the logic is duplicated in tests otherwise"
   [toc nfo]
-  (merge toc nfo))
+  (merge toc nfo (some->> (merge-source-map-lists (extract-source-map-list toc)
```
Comment (torkus) [inline src/strongbox/addon.clj:138]: pick this new map out into a `let` binding

docstr

defn-spec

update tests

File: src/strongbox/addon.clj [hunk 4]
```diff
@@ -501,3 +503,24 @@
   "returns `true` if there is more than 1 release available and the addon isn't pinned"
   [addon map?]
   (releases-available? (update-in addon [:release-list] rest)))
+
+;;
+
+(defn-spec update-nfo! nil?
```
Comment (torkus) [inline src/strongbox/addon.clj:509]: tests

File: src/strongbox/addon.clj [hunk 5]
```diff
@@ -501,3 +503,24 @@
...
+    (nfo/update-nfo install-dir (:dirname grouped-addon) updates)))
+
+(defn-spec switch-source! nil?
+  "switches addon from one source (like curseforge) to another (like wowinterface), rewriting nfo data.
+  `new-source` must appear in the addon's `source-map-list`."
+  [install-dir ::sp/extant-dir, addon :addon/installed, new-source-map :addon/source-map]
+  (when (and (utils/in? new-source-map (:source-map-list addon)) ;; todo: consider moving this check to cli
+             (not (ignored? addon))
+             (not (pinned? addon)))
+    ;; todo: update group-id as well
```
Comment (torkus) [inline src/strongbox/addon.clj:522]: rm

File: src/strongbox/addon.clj [hunk 6]
```diff
@@ -501,3 +503,24 @@
...
+
+;;
+
+(defn-spec update-nfo! nil?
+  "updates the nfo data for the given addon and all of it's grouped addons"
+  [install-dir ::sp/extant-dir, addon :addon/toc+nfo, updates map?]
+  (doseq [grouped-addon (flatten-addon addon)]
+    (nfo/update-nfo install-dir (:dirname grouped-addon) updates)))
+
+(defn-spec switch-source! nil?
```
Comment (torkus) [inline src/strongbox/addon.clj:515]: test

File: src/strongbox/github_api.clj [hunk 7]
```diff
@@ -335,3 +335,9 @@
               (assoc addon :description description)
               addon)))]
     (mapv to-summary result-list)))
+
+;;
+
+(defn-spec make-url (s/or :ok ::sp/url, :error nil?)
```
Comment (torkus) [inline src/strongbox/github_api.clj:341]: docstr

test

nil if no source-id

File: src/strongbox/gitlab_api.clj [hunk 8]
```diff
@@ -265,3 +265,9 @@
                          (clojure.string/split #"/")))] ;; ["group" "owner" "project"]
     (when (> (count bits) 1)
       (clojure.string/join "/" bits))))
+
+;;
+
+(defn-spec make-url (s/or :ok ::sp/url, :error nil?)
```
Comment (torkus) [inline src/strongbox/gitlab_api.clj:271]: docstr

test

nil if no source-id

File: src/strongbox/nfo.clj [hunk 9]
```diff
@@ -54,7 +63,9 @@
 
              ;; where the addon came from and how they identified it
              :source (:source addon)
-             :source-id (:source-id addon)}
+             :source-id (:source-id addon)
+
```
Comment (torkus) [inline src/strongbox/nfo.clj:67]: rm ws

File: src/strongbox/nfo.clj [hunk 10]
```diff
@@ -54,7 +63,9 @@
...
+             :source-map-list [{:source (:source addon) :source-id (:source-id addon)}]}
```
Comment (torkus) [inline src/strongbox/nfo.clj:68]: what happens when we `derive` from derived data? we'd lose whatever `source-map-list` already was ... do we do that anywhere? Is it unreasonable to do so?

File: src/strongbox/nfo.clj [hunk 11]
```diff
@@ -101,12 +100,25 @@
...
-    (read-nfo-file* path opts)))
+                       (when (fs/exists? path)
+                         (warn (format "invalid \"%s\" file, deleting: %s" nfo-filename path))
+                         (rm-nfo-file path)))
+        opts {:no-file? nil
+              :bad-data? bad-data
+              :key-fn keyword
+              :transform-map {:installed-game-track keyword}}
+
+        coerce (fn [nfo-data]
```
Comment (torkus) [inline src/strongbox/nfo.clj:111]: tests if we don't already have some ... I think we do.

File: src/strongbox/nfo.clj (hunk 10 above)
Comment (torkus) [reply src/strongbox/nfo.clj:68]: add note about how nfo only captures sources that have been *installed*, not the list of all known ones that comes from the toc data

File: src/strongbox/specs.clj [hunk 12]
```diff
@@ -274,9 +270,10 @@
                    ::download-count
                    :addon/source
                    :addon/source-id]
-          :opt-un [::description ;; wowinterface summaries have no description
+          :opt-un [::description ;; wowinterface summaries have no description. github may not have a description.
                    :addon/created-date ;; wowinterface summaries have no created date
-                   ::game-track-list ;; more of a set, really
+                   ::game-track-list   ;; more of a set, really
+                   ;;:addon/source-map-list
```
Comment (torkus) [inline src/strongbox/specs.clj:276]: rm

File: src/strongbox/toc.clj [hunk 13]
```diff
@@ -161,6 +161,11 @@
                         {:source "tukui"
                          :source-id x-tukui-id})
 
+         source-map-list (when-let [items (->> [wowi-source curse-source tukui-source]
```
Comment (torkus) [inline src/strongbox/toc.clj:164]: test for this?

File: src/strongbox/ui/cli.clj [hunk 14]
```diff
@@ -629,6 +629,14 @@
   {:browse-local {:label "browse" :value-fn :addon-dir}
    :source {:label "source" :value-fn :source}
    :source-id {:label "ID" :value-fn :source-id}
+   :source-map-list {:label "other sources" :value-fn (fn [row]
+                                                        (->> row
+                                                             :source-map-list
+                                                             (map :source)
+                                                             (remove #(= % (:source row)))
+                                                             vec
```
Comment (torkus) [inline src/strongbox/ui/cli.clj:637]: rm

File: src/strongbox/ui/jfx.clj [hunk 15]
```diff
@@ -1123,6 +1123,21 @@
                                              {:graphic (href-to-hyperlink row)})}
                   :cell-value-factory identity}
          :source-id {:min-width 60 :pref-width 150}
+         :source-map-list {:cell-factory {:fx/cell-type :tree-table-cell
+                                          :describe (fn [row]
+                                                      (let [urls (for [source-map (:source-map-list row)
+                                                                       :let [url (utils/addon-source-map-to-url row source-map)]
```
Comment (torkus) [inline src/strongbox/ui/jfx.clj:1129]: `utils/addon-source-map-to-url` is replicating `*/make-url`logic that already exists

File: src/strongbox/ui/jfx.clj [hunk 16]
```diff
@@ -1424,6 +1439,19 @@
                      (async-handler (juxt (partial cli/set-version addon release) clear-table-selected-items))))
         (:release-list addon)))
 
+(defn-spec build-addon-source-menu (s/or :ok ::sp/list-of-maps, :no-sources nil?)
```
Comment (torkus) [inline src/strongbox/ui/jfx.clj:1442]: docstr

File: src/strongbox/utils.clj [hunk 17]
```diff
@@ -684,6 +684,28 @@
                     nil)
       nil)))
 
+(defn url-encode
+  [string]
+  (if string
+    (java.net.URLEncoder/encode string "UTF-8")
+    ""))
+
+(defn-spec addon-source-map-to-url (s/or :ok ::sp/url, :error nil?)
```
Comment (torkus) [inline src/strongbox/utils.clj:693]: duplicated logic, see `*/make-url` fns

File: test/strongbox/addon_test.clj [hunk 18]
```diff
@@ -533,3 +544,67 @@
...
+                       {:source-map-list [new-source-map]})
+          expected (merge helper/nfo-data
+                          new-source-map
+                          {:source-map-list [{:source "curseforge" :source-id 1} ;; original is preserved
+                                             new-source-map]})] ;; new one is present
+      (helper/install-every-addon!)
+      (is (nil? (addon/switch-source! (helper/install-dir) addon new-source-map)))
+      (is (= expected (nfo/read-nfo (helper/install-dir) (:dirname helper/toc-data)))))))
+
+(deftest merge-toc-nfo--sml1
```
Comment (torkus) [inline test/strongbox/addon_test.clj:594]: `sml1` ? fix

File: test/strongbox/nfo_test.clj [hunk 19]
```diff
@@ -69,20 +71,23 @@
...
+(deftest read-nfo--invalid
   (testing "invalid nfo data returns `nil` and the nfo file is deleted"
     (let [invalid-nfo-data [{} [] 1 {:foo "bar"} "null"]
           expected nil]
       (doseq [nfo-data invalid-nfo-data]
         (spit (utils/join (addon-path) nfo/nfo-filename) (utils/to-json nfo-data))
-        (is (= expected (nfo/read-nfo (install-dir) addon-dir)))
-        (is (not (fs/exists? (nfo/read-nfo (install-dir) addon-dir)))))))
+        (is (= expected (nfo/read-nfo (install-dir) addon-dir)), (str "failed: " nfo-data))))))
+        ;;(is (not (fs/exists? (nfo/read-nfo (install-dir) addon-dir))))))))
```
Comment (torkus) [inline test/strongbox/nfo_test.clj:83]: what is happening here?

File: test/strongbox/nfo_test.clj [hunk 20]
```diff
@@ -207,10 +230,13 @@
...
           updates {:installed-game-track :foo}
 
-          expected nfo-data
+          ;; 2021-12-20: new.
+          source-map-list {:source-map-list [{:source "curseforge" :source-id 321}]}
+
+          expected (merge nfo-data source-map-list)
           expected-log ["new \".strongbox.json\" data is invalid and won't be written to file"]]
 
-      (nfo/write-nfo (install-dir) addon-dir nfo-data)
```
Comment (torkus) [inline test/strongbox/nfo_test.clj:213]: why was this replaced again?

File: test/strongbox/toc_test.clj [hunk 21]
```diff
@@ -89,9 +89,12 @@ SomeAddon.lua")
                      :-toc/game-track :retail
                      :supported-game-tracks [:retail]
                      :installed-version "1.6.1"
-                     ;; wowi is edged out in favour of curseforge unfortunately
+                     ;; wowi is edged out in favour of curseforge ...
                      :source "curseforge"
-                     :source-id 54321}]]
+                     :source-id 54321
+                     ;; ... both are captured here in `:source-map-list` however
```
Comment (torkus) [inline test/strongbox/toc_test.clj:95]: "... however both "

File: src/strongbox/nfo.clj (hunk 10 above)
Comment (torkus) [reply src/strongbox/nfo.clj:68]: the only place `derive` is used is during addon installation. Thereafter `update-nfo` is used. We're safe for now.

**Comment by layday** [comment]
Ooh this is very exciting, I've thought about doing something similar in instawow with the add-ons from the GH catalogue.

**Comment by layday** [comment]
(HNY :)

**Comment by torkus** [comment]
happy new years :)

**Pull request merged**